src/docfx/build/xref/ExternalXrefMapLoader.cs

```diff
@@ -37,47 +41,73 @@ internal class ExternalXrefMapLoader
+                        foreach (var (uid, spec) in externalXrefSpecAndExternalXref.externalXrefSpec)
                         {
                             // for same uid with multiple specs, we should respect the order of the list
-                            result.TryAdd(uid, spec);
+                            externalXrefMap.TryAdd(uid, spec);
                         }
+                        externalXref = externalXrefSpecAndExternalXref.externalXref;
                     }
                 }
 
-                result.TrimExcess();
-                return result;
+                externalXrefMap.TrimExcess();
+                externalXref.TrimExcess();
+
+                return new Lazy<(IReadOnlyDictionary<string, Lazy<ExternalXrefSpec>>, IReadOnlyList<Lazy<ExternalXref>>)>(
+                    () => (externalXrefMap, externalXref));
             }
         }
 
-        public static List<(string, Lazy<ExternalXrefSpec>)> LoadJsonFile(string filePath)
+        public static (List<(string, Lazy<ExternalXrefSpec>)> externalXrefSpec, List<Lazy<ExternalXref>> externalXref) LoadJsonFile(string filePath)
         {
-            var result = new List<(string, Lazy<ExternalXrefSpec>)>();
+            var externalXrefSpec = new List<(string, Lazy<ExternalXrefSpec>)>();
+            var externalXref = new List<Lazy<ExternalXref>>();
+
             var content = File.ReadAllBytes(filePath);
 
             // TODO: cache this position mapping if xref map file not updated, reuse it
-            var (xrefSpecPositions, repositoryUrl) = GetXrefSpecPosAndRepoUrl(content, filePath);
+            var (xrefSpecPositions, xrefPositions, repositoryUrl) = GetXrefSpecPosXrefPosAndRepoUrl(content, filePath);
 
             foreach (var (uid, start, end) in xrefSpecPositions)
             {
-                result.Add((uid, new Lazy<ExternalXrefSpec>(() =>
+                externalXrefSpec.Add((uid, new Lazy<ExternalXrefSpec>(() =>
                 {
                     using var stream = File.Open(filePath, FileMode.Open, FileAccess.Read, FileShare.Read);
                     var spec = JsonUtility.DeserializeData<ExternalXrefSpec>(ReadJsonFragment(stream, start, end), new FilePath(filePath));
                     spec.RepositoryUrl = repositoryUrl;
                     return spec;
                 })));
             }
-            return result;
+
+            foreach (var (start, end) in xrefPositions)
+            {
+                externalXref.Add(new Lazy<ExternalXref>(() =>
+                {
+                    using var stream = File.Open(filePath, FileMode.Open, FileAccess.Read, FileShare.Read);
+                    var xref = JsonUtility.DeserializeData<ExternalXref>(ReadJsonFragment(stream, start, end), new FilePath(filePath));
+                    xref.ReferencedRepositoryUrl = repositoryUrl;
+                    return xref;
+                }));
+            }
+
+            return (externalXrefSpec, externalXref);
```

Comment:
it doesn't make sense to use the same optimization way to load externalXref  as loading externalXrefMap, externalXref should be a really small data size.

Reply:
OK. Yufei has proposed another improvement, we will adjust it in another pr.